### PR TITLE
WIP: Support for Embedded Swift (v2)

### PR DESCRIPTION
**Pull request by sliemeobn**
thanks to @MihaelIsaev for starting the other branch, I was inspired to try on my own by it.

I left String untouched (the entire JS runtime really), but I had to move a few things out of the C lib (ideally we can get rid of the entire thing eventually).

Also, the C header bit field with stuff does not seem to play well with Embedded (do not know why, pushed it into Swift for now)

there is an `Examples/Embedded` demo that shows how it wants to be built.

my hope is that, one day:
1) SwiftPM learns to set the flags on its own with a dedicated "Embedded" mode
2) variadic generics will be supported

for now, a single unsafe flag needs to be added in the package.swift, and bundle resources do not work (SwiftPM will generate code for Foundation bundles).

In general, JavaScriptKit could use a bit of modernization and replace of all the "anys" with variadic generics.
Also, I don't see the value of using class hierarchies (does not play well with embedded) - but I tried to leave as much untouched as possible.

Promises, throwing functions, and typed arrays are left unsupported entirely for now, the rest should work with almost no API surface change.

### CODE REVIEW

**Comment by sliemeobn** [comment]
also @MihaelIsaev, I think your Foundation package is a great idea and a great start. 
for whatever reason I ended up needing `strlen` as well, and `memmove` is needed to get string interpolation going.

do you think you could add these from the wasi/musl libc as well, and add proper license attribution to your package?
also, IMHO the name "Foundation" is very confusing. maybe we can come up with a better name?

File: Sources/JavaScriptKit/FundamentalObjects/JSObject.swift
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/FundamentalObjects/JSObject.swift:234]: Could you elaborate on why this protocol is needed in its doc comment?

File: Sources/JavaScriptKit/FundamentalObjects/JSSymbol.swift
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/FundamentalObjects/JSSymbol.swift:13]: Can we unconditionally use `description.jsValue` here?

File: Examples/Embedded/Sources/EmbeddedApp/_thingsThatShouldNotBeNeeded.swift
Comment (kateinoigakukun) [inline Examples/Embedded/Sources/EmbeddedApp/_thingsThatShouldNotBeNeeded.swift:4]: Could you try adding `@_used` (for Swift decls) or `__attribute__((used))` (for C decls) to teach LTO not to DCE those decls?

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:240]: There is no reason using `cdecl` instead of `silgen_name` here, so I think we can use `silgen_name` and call those functions with Swift calling convention. I'll address this after merging your change, so don't care about this comment, just a note for me.

**Comment by kateinoigakukun** [comment]
I think it's a much mergeable approach. Happy to merge once CI passed and a few my concerns are resolved. Thank you for tackling on this!

**Comment by sliemeobn** [comment]
thanks for reviewing, I'll try address the comments later today, they all makes sense.

the biggest issue with 5.10 compatibility I have right now is that the swift compiler seems to freak out about `@_expose(wasm)` even when it is `#if`ed away.

I'll try my luck with `@_used` on the cdecls later, maybe we stick to the C shims for exposing to wasm untouched for now...

File: Examples/Embedded/Sources/EmbeddedApp/_thingsThatShouldNotBeNeeded.swift
Comment (sliemeobn) [reply Examples/Embedded/Sources/EmbeddedApp/_thingsThatShouldNotBeNeeded.swift:4]: I tried `@_used` for both `@_expose` and `@_cdecl`, does not work : (
I'll post in the forums about it, maybe this can be fixed in the future somehow.
An underscored public is is the best I can come up with atm.

File: Sources/JavaScriptKit/FundamentalObjects/JSFunction.swift
Comment (sliemeobn) [reply Sources/JavaScriptKit/FundamentalObjects/JSFunction.swift:90]: I shied away from it not to add potentially conflicting overloads.
But this should be fine to add.

File: Sources/JavaScriptKit/FundamentalObjects/JSSymbol.swift
Comment (sliemeobn) [reply Sources/JavaScriptKit/FundamentalObjects/JSSymbol.swift:13]: we sure can

File: Sources/JavaScriptKit/FundamentalObjects/JSFunction.swift
Comment (sliemeobn) [reply Sources/JavaScriptKit/FundamentalObjects/JSFunction.swift:247]: just to clarify, do you mean to use the swift-based "decoding" as I wrote here for embedded in general? (ie: just pass an "unsigned short" through the C shim?)

**Comment by sliemeobn** [comment]
@kateinoigakukun with CI passing, are there any issues left you'd want me to address before considering a merge?

**Comment by kateinoigakukun** [comment]
Sorry I just missed the CI status update

**Review by kateinoigakukun — APPROVED** [review]

**Pull request merged**
